cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,223 @@
+
+            # copy smaller component into larger one.
+            c1_order = (
+                reversed(circuit_graph[c1])
+                if circuit_graph[c1][-1] == q1
+                else iter(circuit_graph[c1])
+            )
+            for q in c1_order:
+                if circuit_graph[c0][0] == q0:
+                    circuit_graph[c0].appendleft(q)
+                else:
+                    circuit_graph[c0].append(q)
+                component_id[q] = c0
+
+        graph = sorted(
+            [circuit_graph[c] for c in set(component_id.values())], key=len, reverse=True
+        )
+        return graph, partners
+
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps disjoint lines of logical qubits onto lines of physical qubits.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        mapped_physicals: Set['cirq.Qid'] = set()
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph, partners = self._make_circuit_graph(circuit)
+
+        def next_physical(
+            current_physical: 'cirq.Qid', partner: 'cirq.Qid', isolated: bool = False
+        ) -> 'cirq.Qid':
+            # Greedily map to highest degree neighbor that is available
+            if not isolated:
+                sorted_neighbors = sorted(
+                    self.device_graph.neighbors(current_physical),
+                    key=lambda x: self.device_graph.degree(x),
+                    reverse=True,
+                )
+                for neighbor in sorted_neighbors:
+                    if neighbor not in mapped_physicals:
+                        return neighbor
+                # If cannot map onto one long line of physical qubits, then break down into multiple
+                # small lines by finding nearest available qubit to the physical center
+            return self._closest_unmapped_qubit(partner, mapped_physicals)
+
+        pq = self.center
+        for i, logical_line in enumerate(circuit_graph):
+            for j, lq in enumerate(logical_line):
+                mapped_physicals.add(pq)
+                qubit_map[lq] = pq
+
+                if j < len(logical_line) - 1:
+                    pq = next_physical(pq, self.center)
+
+                # Edge case: if mapping n qubits on an n-qubit device should not call next_physical
+                # when finished mapping the last logical qubit else will raise an error.
+                elif i < len(circuit_graph) - 1:
+                    if len(circuit_graph[i + 1]) > 1:
+                        pq = next_physical(pq, self.center)
+                    else:
+                        partner = qubit_map[partners[lq]] if lq in partners else self.center
+                        pq = next_physical(pq, partner, isolated=True)
```

Comment:
Let's clean this up to call `next_physical` n times instead of n+1 times to avoid edge cases.